sklearn_numba_dpex/common/topk.py

```diff
@@ -0,0 +1,692 @@
+# This implementation takes inspiration from other open-source implementations, such as
+# [1] https://github.com/pytorch/pytorch/blob/master/caffe2/operators/\
+#    top_k_radix_selection.cuh
+# or
+# [2] https://developer.nvidia.com/blog/gpu-pro-tip-fast-histograms-using-shared\
+# -atomics-maxwell
+
+# TODO: apply Dr. TopK optimizations
+
+import math
+from functools import lru_cache
+
+import dpctl.tensor as dpt
+import numba_dpex as dpex
+import numpy as np
+
+from sklearn_numba_dpex.common._utils import (
+    _get_sequential_processing_device,
+    check_power_of_2,
+)
+from sklearn_numba_dpex.common.kernels import (
+    make_initialize_to_zeros_2d_kernel,
+    make_sum_reduction_2d_kernel,
+)
+
+zero_idx = np.int64(0)
+count_one_as_a_long = np.int64(1)
+count_one_as_an_int = np.int32(1)
+
+
+def topk(array_in, k, group_sizes=None):
```

Comment:
insist on the non-deterministic results in the docstring: the garantee to return valid topks, but not necessarily the same if the threshold value have several occurences